From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: zhangyuheng <zhangyuheng@lunadeer.cn>
Date: Tue, 11 Mar 2025 13:50:42 +0800
Subject: [PATCH] Allow sand duplication


diff --git a/net/minecraft/world/entity/Entity.java b/net/minecraft/world/entity/Entity.java
index 2cd2ce1060f567be6c72b7bc9d02651ec7166203..8788722012311737eaef7dada76240316ef6fa5d 100644
--- a/net/minecraft/world/entity/Entity.java
+++ b/net/minecraft/world/entity/Entity.java
@@ -4245,13 +4245,18 @@ public abstract class Entity implements SyncedDataHolder, Nameable, EntityAccess
                         targetPos, 16, // load 16 blocks to be safe from block physics
                         ca.spottedleaf.concurrentutil.util.Priority.HIGH,
                         (chunks) -> {
-                            net.minecraft.world.level.levelgen.feature.EndPlatformFeature.createEndPlatform(destination, targetPos.below(), true, null);
+                            // DeerFolia start - Vanilla end teleportation
+                            // net.minecraft.world.level.levelgen.feature.EndPlatformFeature.createEndPlatform(destination, targetPos.below(), true, null);
+                            Vec3 finalPos;
+                            if (this instanceof Player) finalPos = Vec3.atBottomCenterOf(targetPos.below());
+                            else finalPos = Vec3.atBottomCenterOf(targetPos);
+                            // DeerFolia end
 
                             // the portal obsidian is placed at targetPos.y - 2, so if we want to place the entity
                             // on the obsidian, we need to spawn at targetPos.y - 1
                             portalInfoCompletable.complete(
                                 new net.minecraft.world.level.portal.TeleportTransition(
-                                    destination, Vec3.atBottomCenterOf(targetPos.below()), Vec3.ZERO, Direction.WEST.toYRot(), 0.0f,
+                                    destination, finalPos, this.getDeltaMovement(), 90.0f, 0.0f,    // DeerFolia Vanilla end teleportation
                                     Relative.union(Relative.DELTA, Set.of(Relative.X_ROT)),
                                     TeleportTransition.PLAY_PORTAL_SOUND.then(TeleportTransition.PLACE_PORTAL_TICKET),
                                     org.bukkit.event.player.PlayerTeleportEvent.TeleportCause.END_PORTAL
@@ -4449,6 +4454,11 @@ public abstract class Entity implements SyncedDataHolder, Nameable, EntityAccess
         if (!this.canPortalAsync(destination, takePassengers)) {
             return false;
         }
+        // DeerFolia start - sync end platform spawning & entity teleportation
+        final java.util.function.Consumer<Entity> tpComplete = type == PortalType.END && destination.getTypeKey() == net.minecraft.world.level.dimension.LevelStem.END
+                ? e -> net.minecraft.world.level.levelgen.feature.EndPlatformFeature.createEndPlatform(destination, ServerLevel.END_SPAWN_POINT.below(), true)
+                : teleportComplete;
+        // DeerFolia end - sync end platform spawning & entity teleportation
 
         Vec3 initialPosition = this.position();
         ChunkPos initialPositionChunk = new ChunkPos(
@@ -4513,9 +4523,14 @@ public abstract class Entity implements SyncedDataHolder, Nameable, EntityAccess
                         info.postTeleportTransition().onTransition(teleported);
                     }
 
-                    if (teleportComplete != null) {
-                        teleportComplete.accept(teleported);
+                    // DeerFolia start - sync end platform spawning & entity teleportation
+                    // if (teleportComplete != null) {
+                    //     teleportComplete.accept(teleported);
+                    // }
+                    if (tpComplete != null) {
+                        tpComplete.accept(teleported);
                     }
+                    // DeerFolia end - sync end platform spawning & entity teleportation
                 }
             );
         });
diff --git a/net/minecraft/world/entity/item/FallingBlockEntity.java b/net/minecraft/world/entity/item/FallingBlockEntity.java
index 1fa5e6a12b943e889bde566038a632a6adcf319e..6b8727c1b123e2ca85d97ea104ab6626d3a697e3 100644
--- a/net/minecraft/world/entity/item/FallingBlockEntity.java
+++ b/net/minecraft/world/entity/item/FallingBlockEntity.java
@@ -65,7 +65,7 @@ public class FallingBlockEntity extends Entity {
     public float fallDamagePerDistance;
     @Nullable
     public CompoundTag blockData;
-    public boolean forceTickAfterTeleportToDuplicate;
+    public boolean forceTickAfterTeleportToDuplicate = true;    // DeerFolia - Sand duplication
     protected static final EntityDataAccessor<BlockPos> DATA_START_POS = SynchedEntityData.defineId(FallingBlockEntity.class, EntityDataSerializers.BLOCK_POS);
     public boolean autoExpire = true; // Paper - Expand FallingBlock API
 
diff --git a/net/minecraft/world/level/block/EndPortalBlock.java b/net/minecraft/world/level/block/EndPortalBlock.java
index 554d75ac1374d7d93977a10e06fcf51259830c97..a5285e9afea331e8a7526e5f3876f3ff8395915b 100644
--- a/net/minecraft/world/level/block/EndPortalBlock.java
+++ b/net/minecraft/world/level/block/EndPortalBlock.java
@@ -67,6 +67,11 @@ public class EndPortalBlock extends BaseEntityBlock implements Portal {
                 if (level.paperConfig().misc.disableEndCredits) {serverPlayer.seenCredits = true; return;} // Paper - Option to disable end credits
                 serverPlayer.showEndCredits();
             } else {
+                // DeerFolia start - unsafe teleportation
+                if (!(entity instanceof net.minecraft.world.entity.player.Player)) {
+                    entity.endPortalLogicAsync(pos);
+                }
+                // DeerFolia end
                 entity.setAsInsidePortal(this, pos);
             }
         }
